### ISSUE TITLE
Wrong diagnostic when returning `_` with a lifetime

### ISSUE THREAD

**Issue by dev-ardi** [rust-lang/rust#125488]
The code
```rs
struct S<'a>(&'a ());
fn f(s: S<'_>) -> _ {
    s
}
```
Emits this diagnostic, which is wrong in this case.
```rs
error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
 --> src/lib.rs:2:19
  |
2 | fn f(s: S<'_>) -> _ {
  |                   ^
  |                   |
  |                   not allowed in type signatures
  |                   help: replace with the correct return type: `S<'static>`

For more information about this error, try `rustc --explain E0121`.
```
It should say `help: replace with the correct return type: S<'_>`
cc: https://github.com/rust-lang/rust-analyzer/issues/16676

**Comment by jieyouxu** [comment]
Cannot reproduce: the code snippet as given produces

```
warning: struct `S` is never constructed
 --> src/lib.rs:1:8
  |
1 | struct S<'a>(&'a ());
  |        ^
  |
  = note: `#[warn(dead_code)]` on by default

warning: function `f` is never used
 --> src/lib.rs:2:4
  |
2 | fn f(s: S<'_>) -> S<'_> {
  |    ^

warning: `playground` (lib) generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.36s
```

on stable 1.78.0

**Comment by jieyouxu** [comment]
Did you mean to write this code snippet instead?

```
struct S<'a>(&'a ());
fn f(s: S<'_>) -> _ {
    s
}
```
